### PR TITLE
Fix SDP m= line parser, verify port & port count are numeric

### LINKED ISSUE #3366: Segmentation fault upon invalid READ access
### Describe the bug

The issue is a segmentation fault in the `create_answer()` function in sipecho.c caused by an invalid read access of a null pointer. 

### Steps to reproduce

1. Compile PJSIP binaries under pjproject directory: .`/configure --disable-strip CFLAGS=-g && make dep && make clean && make`

2. Compile Baresip binary under baresip directory: `make -C deps/re install PREFIX=$PWD/bin && make -C deps/rem install PREFIX=$PWD/bin LIBRE_MK=$SIP_FUZZER_DIR/baresip/deps/re/mk/re.mk && make PREFIX=$PWD/bin LIBRE_MK=deps/re/mk/re.mk LIBRE_INC=bin/include/re/ LIBRE_SO=bin/lib/ LIBREM_PATH=bin/ install`

3. Launch SIP Echo service: .`/pjproject/pjsip-apps/bin/samples/x86_64-unknown-linux-gnu/sipecho`

4. Launch Baresip client: `./baresip -f docs/examples -e \"d test@127.0.0.1:5060\" -n 127.0.0.1"
`
5. Baresip sends the following INVITE message to the SIP Echo application:

```
	INVITE sip:test@127.0.0.1:5060 SIP/2.0\r\n
	Via: SIP/2.0/UDP 127.0.0.1:41937;branch=z9hG4bK27e3f61708de5539;rport\r\n
	Contact: <sip:user-0x805010@127.0.0.1:41937>\r\n
	Max-Forwards: 70\r\nTo: <sip:test@127.0.0.1:5060>\r\n
	From: <sip:user@example.com>;tag=0435ab00f8256ef0\r\n
	Call-ID: 5bc68ea866c3ccf0\r\n
	CSeq: 4261 INVITE\r\n
	User-Agent: baresip v0.6.2 (x86_64/linux)\r\n
	Allow: INVITE,ACK,BYE,CANCEL,OPTIONS,NOTIFY,SUBSCRIBE,INFO,MESSAGE,REFER\r\n
	Supported: gruu\r\n
	Content-Type: application/sdp\r\n
	Content-Length: 411\r\n\r\n
	v=0\r\n
	o=- 4179833163 558260718 IN IP4 127.0.0.1\r\n
	s=-\r\n
	c=IN IP4 127.0.0.1\r\n
	t=0 0\r\n
	a=tool:baresip 0.6.2\r\n
	m=audio 46640 RTP/AVP 0 8 101\r\n
	a=rtpmap:0 PCMU/8000\r\n
	a=rtpmap:8 PCMA/8000\r\n
	a=rtpmap:101 telephone-event/8000\r\n
	a=fmtp:101 0-15\r\n
	a=sendrecv\r\n
	a=label:1\r\n
	a=rtcp-rsize\r\n
	a=ssrc:3089433414 cname:sip:user@example.com\r\n
	a=ptime:20\r\n
	a=ebuacip:version 0\r\n
	a=ebuacip:jb 0\r\n
	a=ebuacip:jbdef 0 fixed 200\r\n
	a=ebuacip:qosrec 46\r\n
```

6. SIP Echo application hits the segmentation fault.

### PJSIP version

2.13, 2.8

### Context

The cause of this SEGV is a READ memory access to a zero page when PJSIP receives the incoming INVITE message and creates an echoing response based on this offer message. Specifically, the trigger is that the SDP connection variable `c` is a NULL pointer. Logging its attributes such as `c->addr.slen` with `PJ_LOG` in sipecho.c:374 is an invalid READ of NULL address, raising the SEGV signal.

This finding indicates that the SIP Echo service's source code has flaws. The echo message's connection is assigned (i.e., executing `c = m->conn ? m->conn : answer->conn` in sipecho.c:371) before the SDP parser scans the connection attribute in the SDP body (i.e., calling pjmedia_sdp_parse() function).

To address this issue, we recommend implementing a status check in the `create_answer()` function. We have included a patch for this proposed fix and would appreciate your review. You can also find the log after the application of this patch below. Thank you!

[proposed_fix.txt](https://github.com/pjsip/pjproject/files/10669181/proposed_fix.txt)
(There is something wrong with submitting .patch file on GitHub. I apologize for this inconvenience.) 

### Log, call stack, etc

```shell
ASAN:DEADLYSIGNAL
=================================================================
==11==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000028 (pc 0x0000004f18ba bp 0x61b0000000a8 sp 0x7f228b4eb5d0 T1)
==11==The signal is caused by a READ memory access.
==11==Hint: address points to the zero page.
    #0 0x4f18b9 in create_answer /root/server-afl/programs/sip/pjproject/pjsip-apps/build/../src/samples/sipecho.c:374:2
    #1 0x4f29cc in on_rx_request /root/server-afl/programs/sip/pjproject/pjsip-apps/build/../src/samples/sipecho.c:505:11
    #2 0x553e11 in pjsip_endpt_process_rx_data /root/server-afl/programs/sip/pjproject/pjsip/build/../src/pjsip/sip_endpoint.c:893:13
    #3 0x55212a in endpt_on_rx_msg /root/server-afl/programs/sip/pjproject/pjsip/build/../src/pjsip/sip_endpoint.c:1043:5
    #4 0x56f9a9 in pjsip_tpmgr_receive_packet /root/server-afl/programs/sip/pjproject/pjsip/build/../src/pjsip/sip_transport.c
    #5 0x577069 in udp_on_read_complete /root/server-afl/programs/sip/pjproject/pjsip/build/../src/pjsip/sip_transport_udp.c:192:3
    #6 0x5f6394 in ioqueue_dispatch_read_event /root/server-afl/programs/sip/pjproject/pjlib/build/../src/pj/ioqueue_common_abs.c:605:6
    #7 0x5fdd9d in pj_ioqueue_poll /root/server-afl/programs/sip/pjproject/pjlib/build/../src/pj/ioqueue_select.c:1070:7
    #8 0x55328c in pjsip_endpt_handle_events2 /root/server-afl/programs/sip/pjproject/pjsip/build/../src/pjsip/sip_endpoint.c:744:6
    #9 0x4f0c3e in worker_proc /root/server-afl/programs/sip/pjproject/pjsip-apps/build/../src/samples/sipecho.c:167:2
    #10 0x600c1e in thread_main /root/server-afl/programs/sip/pjproject/pjlib/build/../src/pj/os_core_unix.c:541:27
    #11 0x7f228e9af6b9 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76b9)
    #12 0x7f228e2cb51c in clone (/lib/x86_64-linux-gnu/libc.so.6+0x10751c)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /root/server-afl/programs/sip/pjproject/pjsip-apps/build/../src/samples/sipecho.c:374:2 in create_answer
Thread T1 created by T0 here:
    #0 0x4aa696 in pthread_create /scratch/llvm/clang-4/xenial/final/llvm.src/projects/compiler-rt/lib/asan/asan_interceptors.cc:305:3
    #1 0x6005b5 in pj_thread_create /root/server-afl/programs/sip/pjproject/pjlib/build/../src/pj/os_core_unix.c:634:10
    #2 0x4ef3ff in init_stack /root/server-afl/programs/sip/pjproject/pjsip-apps/build/../src/samples/sipecho.c:289:14
    #3 0x4ef3ff in main /root/server-afl/programs/sip/pjproject/pjsip-apps/build/../src/samples/sipecho.c:607
    #4 0x7f228e1e483f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2083f)

==11==ABORTING

************************************Log Before Patch************************************
Ready for incoming calls. Press ENTER to quit...

Menu:
  h    Hangup all calls
  l    Enable message logging
  q    Quit

************************************Log After Patch************************************
Ready for incoming calls. Press ENTER to quit...

Menu:
  h    Hangup all calls
  l    Enable message logging
  q    Quit

14:28:48.373              sipecho.c !.RX INVITE from 127.0.0.1
14:28:48.874              sipecho.c  .RX INVITE from 127.0.0.1
14:28:48.874              sipecho.c  .Call 0: creating answer:
answer's media count: 1
14:28:48.879              sipecho.c  ..Call 0: state changed to CONFIRMED
14:28:48.880              sipecho.c  .....Call 0: DISCONNECTED [reason=200 (OK)]
```

### PR DESCRIPTION
**Pull request by nanangizz**
To fix #3366.

From the provided pcap, the offending SDP is:
```
v=0
o=- 3973070478 208142209 IN IP4 127.0.0.1
s=-ÿc=IN IP4 127.0.0.1
t=0 0
a=tool:baresip 0.6.2
m=auÿÿÿ34718 RTP/AVP 0 8 101
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-15
a=sendrecv
a=label:1
a=rtcp-rsize
a=ptime:20
```
Note that there is no c= line.

The problem starts from the m= line here:
```
m=auÿÿÿ34718 RTP/AVP 0 8 101
```
the SDP parser will treat string `RTP/AVP` as port and it will be translated to `0` by `pj_strtoul()`. When the port is `0`, no connection line in both session level & media level will be tolerated, so the SDP is considered valid and eventually it may introduce issues in app (as reported). This PR will make sure that the port & port count fields are numeric, so above m= line will be considered invalid.

### CODE REVIEW

**Review by bennylp — COMMENTED** [review]
While we are here, maybe fix other pj_strtoul() usages as well?

**Comment by nanangizz** [comment]
> While we are here, maybe fix other pj_strtoul() usages as well?

hmm.. will create another PR for this.

File: pjmedia/src/pjmedia/sdp.c [hunk 1]
```diff
@@ -1165,22 +1167,30 @@ static void parse_media(pj_scanner *scanner, pjmedia_sdp_media *med,
...
     /* port */
     pj_scan_get(scanner, &cs_token, &str);
-    med->desc.port = (unsigned short)pj_strtoul(&str);
-    if (pj_scan_is_eof(scanner)) {
+    status = pj_strtoul3(&str, &num, 10);
+    if (status != PJ_SUCCESS || pj_scan_is_eof(scanner)) {
         on_scanner_error(scanner);
         return;
     }
+    med->desc.port = (unsigned short)num;
```
Comment (sauwming) [inline pjmedia/src/pjmedia/sdp.c:1183]: do we need to guard against overflow?
also for the port number below.

Comment (nanangizz) [reply pjmedia/src/pjmedia/sdp.c:1183]: IMHO ideally yes, but not sure if it should be done in that parsing function as the SDP RFC does not seem to specify the limit. But our port data type is uint16.

**Comment by sauwming** [comment]
Perhaps we can also add the problematic SDP into our seed corpus?
https://google.github.io/oss-fuzz/advanced-topics/ideal-integration/#seed-corpus
```
For better OSS-Fuzz integration, the seed corpus should be available in revision
control (it can be the same as or different from the source code).
It should be regularly extended with the inputs that (used to) trigger bugs
and/or touch new parts of the code.
```

**Review by bennylp — APPROVED** [review]

**Comment by nanangizz** [comment]
> Perhaps we can also add the problematic SDP into our seed corpus? https://google.github.io/oss-fuzz/advanced-topics/ideal-integration/#seed-corpus
> 
> ```
> For better OSS-Fuzz integration, the seed corpus should be available in revision
> control (it can be the same as or different from the source code).
> It should be regularly extended with the inputs that (used to) trigger bugs
> and/or touch new parts of the code.
> ```

Just did, but I haven't tested it, don't know how :)

**Comment by sauwming** [comment]
Need to export ` LIB_FUZZING_ENGINE` and `CFLAGS` as described in `Fuzzing.md`.

**Review by sauwming — APPROVED** [review]

**Comment by nanangizz** [comment]
The test built fine and run fine (perhaps :D at least it does not reject the new input), because after running for a quite long time it eventually printed:
```
==13578==AddressSanitizer: Thread limit (4194304 threads) exceeded. Dying.
```

**Pull request merged**